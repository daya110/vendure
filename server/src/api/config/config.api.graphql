type Query {
  config: Config!
}

type Config {
    customFields: JSON
}
