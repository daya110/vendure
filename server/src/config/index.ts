export * from './asset-naming-strategy/asset-naming-strategy';
export * from './asset-naming-strategy/default-asset-naming-strategy';
export * from './asset-preview-strategy/asset-preview-strategy';
export * from './asset-storage-strategy/asset-storage-strategy';
export * from './email/email-options';
export * from './email/email-transport-options';
export * from './entity-id-strategy/auto-increment-id-strategy';
export * from './entity-id-strategy/entity-id-strategy';
export * from './entity-id-strategy/uuid-id-strategy';
export * from './order-merge-strategy/order-merge-strategy';
export * from './payment-method/fakepal-payment-method-config';
export * from './payment-method/gripe-payment-method-config';
export * from './payment-method/payment-method-handler';
export * from './promotion/default-promotion-actions';
export * from './promotion/default-promotion-conditions';
export * from './promotion/promotion-action';
export * from './promotion/promotion-condition';
export * from './shipping-method/default-shipping-calculator';
export * from './shipping-method/default-shipping-eligibility-checker';
export * from './shipping-method/shipping-calculator';
export * from './shipping-method/shipping-eligibility-checker';
export * from './vendure-config';
export * from './vendure-plugin/vendure-plugin';
