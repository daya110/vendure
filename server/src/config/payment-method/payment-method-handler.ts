import { ConfigArg } from 'shared/generated-types';

import { StateMachineConfig } from '../../common/finite-state-machine';
import { Order } from '../../entity/order/order.entity';
import { PaymentMetadata } from '../../entity/payment/payment.entity';
import {
    PaymentState,
    PaymentTransitionData,
} from '../../service/helpers/payment-state-machine/payment-state';
import { argsArrayToHash, ConfigArgs, ConfigArgValues } from '../common/config-args';

export type PaymentMethodArgType = 'int' | 'string' | 'boolean';
export type PaymentMethodArgs = ConfigArgs<PaymentMethodArgType>;
export type OnTransitionStartReturnType = ReturnType<Required<StateMachineConfig<any>>['onTransitionStart']>;

export type OnTransitionStartFn<T extends PaymentMethodArgs> = (
    fromState: PaymentState,
    toState: PaymentState,
    args: ConfigArgValues<T>,
    data: PaymentTransitionData,
) => OnTransitionStartReturnType;

export interface PaymentConfig {
    amount: number;
    state: PaymentState;
    transactionId?: string;
    metadata?: PaymentMetadata;
}

export type CreatePaymentFn<T extends PaymentMethodArgs> = (
    order: Order,
    args: ConfigArgValues<T>,
    metadata: PaymentMetadata,
) => PaymentConfig | Promise<PaymentConfig>;

export interface PaymentMethodConfigOptions<T extends PaymentMethodArgs = {}> {
    code: string;
    name: string;
    createPayment: CreatePaymentFn<T>;
    args: T;
    onStateTransitionStart?: OnTransitionStartFn<T>;
}

/**
 * A PaymentMethodHandler contains the code which is used to generate a Payment when a call to the
 * `addPaymentToOrder` mutation is made. If contains any necessary steps of interfacing with a
 * third-party payment gateway before the Payment is created and can also define actions to fire
 * when the state of the payment is changed.
 */
export class PaymentMethodHandler<T extends PaymentMethodArgs = {}> {
    readonly code: string;
    readonly name: string;
    readonly args: T;
    private readonly createPaymentFn: CreatePaymentFn<T>;
    private readonly onTransitionStartFn?: OnTransitionStartFn<T>;

    constructor(config: PaymentMethodConfigOptions<T>) {
        this.code = config.code;
        this.name = config.name;
        this.args = config.args;
        this.createPaymentFn = config.createPayment;
        this.onTransitionStartFn = config.onStateTransitionStart;
    }

    async createPayment(order: Order, args: ConfigArg[], metadata: PaymentMetadata) {
        const paymentConfig = await this.createPaymentFn(order, argsArrayToHash(args), metadata);
        return {
            method: this.code,
            ...paymentConfig,
        };
    }

    onStateTransitionStart(
        fromState: PaymentState,
        toState: PaymentState,
        args: ConfigArg[],
        data: PaymentTransitionData,
    ): OnTransitionStartReturnType {
        if (typeof this.onTransitionStartFn === 'function') {
            return this.onTransitionStartFn(fromState, toState, argsArrayToHash(args), data);
        } else {
            return true;
        }
    }
}
