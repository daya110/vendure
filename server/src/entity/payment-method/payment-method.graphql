type PaymentMethod implements Node {
    id: ID!
    createdAt: DateTime!
    updatedAt: DateTime!
    code: String!
    enabled: Boolean!
    configArgs: [ConfigArg!]!
}

input UpdatePaymentMethodInput {
    id: ID!
    code: String
    enabled: Boolean
    configArgs: [ConfigArgInput!]
}
