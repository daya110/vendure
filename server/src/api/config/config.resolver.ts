import { Query, Resolver } from '@nestjs/graphql';

import { getConfig, VendureConfig } from '../../config/vendure-config';
import { ConfigService } from '../../service/config.service';

@Resolver('Config')
export class ConfigResolver {
    constructor(private configService: ConfigService) {}

    /**
     * Exposes a subset of the VendureConfig which may be of use to clients.
     */
    @Query()
    config(): Partial<VendureConfig> {
        return {
            customFields: this.configService.customFields,
        };
    }
}
