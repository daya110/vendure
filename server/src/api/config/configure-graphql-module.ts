import { DynamicModule } from '@nestjs/common';
import { GqlModuleOptions, GraphQLModule, GraphQLTypesLoader } from '@nestjs/graphql';
import { GraphQLUpload } from 'apollo-server-core';
import { extendSchema, printSchema } from 'graphql';
import { GraphQLDateTime } from 'graphql-iso-date';
import GraphQLJSON from 'graphql-type-json';

import { ConfigModule } from '../../config/config.module';
import { ConfigService } from '../../config/config.service';
import { I18nModule } from '../../i18n/i18n.module';
import { I18nService } from '../../i18n/i18n.service';
import { getPluginAPIExtensions } from '../../plugin/plugin-utils';
import { ApiSharedModule } from '../api-internal-modules';
import { IdCodecService } from '../common/id-codec.service';
import { IdEncoderExtension } from '../middleware/id-encoder-extension';
import { TranslateErrorExtension } from '../middleware/translate-errors-extension';

import { generateListOptions } from './generate-list-options';
import { addGraphQLCustomFields } from './graphql-custom-fields';

export interface GraphQLApiOptions {
    apiType: 'shop' | 'admin';
    typePaths: string[];
    apiPath: string;
    // tslint:disable-next-line:ban-types
    resolverModule: Function;
}

/**
 * Dynamically generates a GraphQLModule according to the given config options.
 */
export function configureGraphQLModule(
    getOptions: (configService: ConfigService) => GraphQLApiOptions,
): DynamicModule {
    return GraphQLModule.forRootAsync({
        useFactory: (
            configService: ConfigService,
            i18nService: I18nService,
            idCodecService: IdCodecService,
            typesLoader: GraphQLTypesLoader,
        ) => {
            return createGraphQLOptions(
                i18nService,
                configService,
                idCodecService,
                typesLoader,
                getOptions(configService),
            );
        },
        inject: [ConfigService, I18nService, IdCodecService, GraphQLTypesLoader],
        imports: [ConfigModule, I18nModule, ApiSharedModule],
    });
}

function createGraphQLOptions(
    i18nService: I18nService,
    configService: ConfigService,
    idCodecService: IdCodecService,
    typesLoader: GraphQLTypesLoader,
    options: GraphQLApiOptions,
): GqlModuleOptions {
    // Prevent `Type "Node" is missing a "resolveType" resolver.` warnings.
    // See https://github.com/apollographql/apollo-server/issues/1075
    const dummyResolveType = {
        __resolveType() {
            return null;
        },
    };

    return {
        path: '/' + options.apiPath,
        typeDefs: createTypeDefs(options.apiType),
        include: [options.resolverModule],
        resolvers: {
            JSON: GraphQLJSON,
            DateTime: GraphQLDateTime,
            Node: dummyResolveType,
            PaginatedList: dummyResolveType,
            Upload: GraphQLUpload || dummyResolveType,
            SearchResultPrice: {
                __resolveType(value: any) {
                    return value.hasOwnProperty('value') ? 'SinglePrice' : 'PriceRange';
                },
            },
        },
        uploads: {
            maxFileSize: configService.assetOptions.uploadMaxFileSize,
        },
        playground: true,
        debug: true,
        context: (req: any) => req,
        extensions: [
            () => new TranslateErrorExtension(i18nService),
            () => new IdEncoderExtension(idCodecService),
        ],
        // This is handled by the Express cors plugin
        cors: false,
    };

    /**
     * Generates the server's GraphQL schema by combining:
     * 1. the default schema as defined in the source .graphql files specified by `typePaths`
     * 2. any custom fields defined in the config
     * 3. any schema extensions defined by plugins
     */
    function createTypeDefs(apiType: 'shop' | 'admin'): string {
        const customFields = configService.customFields;
        const typeDefs = typesLoader.mergeTypesByPaths(...options.typePaths);
        let schema = generateListOptions(typeDefs);
        schema = addGraphQLCustomFields(schema, customFields);
        const pluginSchemaExtensions = getPluginAPIExtensions(configService.plugins, apiType).map(
            e => e.schema,
        );

        for (const documentNode of pluginSchemaExtensions) {
            schema = extendSchema(schema, documentNode);
        }
        return printSchema(schema);
    }
}
