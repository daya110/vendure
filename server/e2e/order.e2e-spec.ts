import gql from 'graphql-tag';
import { CreateAddressInput, GetCustomerList } from 'shared/generated-types';

import { GET_CUSTOMER_LIST } from '../../admin-ui/src/app/data/definitions/customer-definitions';
import { PaymentMethodHandler } from '../src/config/payment-method/payment-method-handler';

import { TEST_SETUP_TIMEOUT_MS } from './config/test-config';
import { TestClient } from './test-client';
import { TestServer } from './test-server';

describe('Orders', () => {
    const client = new TestClient();
    const server = new TestServer();

    beforeAll(async () => {
        const token = await server.init(
            {
                productCount: 10,
                customerCount: 2,
            },
            {
                paymentOptions: {
                    paymentMethodHandlers: [testPaymentMethod, testFailingPaymentMethod],
                },
            },
        );
        await client.init();
    }, TEST_SETUP_TIMEOUT_MS);

    afterAll(async () => {
        await server.destroy();
    });

    describe('as anonymous user', () => {
        let firstOrderItemId: string;
        let createdCustomerId: string;

        beforeAll(async () => {
            await client.asAnonymousUser();
        });

        it('addItemToOrder starts with no session token', () => {
            expect(client.getAuthToken()).toBe('');
        });

        it('activeOrder returns null before any items have been added', async () => {
            const result = await client.query(GET_ACTIVE_ORDER);
            expect(result.activeOrder).toBeNull();
        });

        it('activeOrder creates an anonymous session', () => {
            expect(client.getAuthToken()).not.toBe('');
        });

        it('addItemToOrder creates a new Order with an item', async () => {
            const result = await client.query(ADD_ITEM_TO_ORDER, {
                productVariantId: 'T_1',
                quantity: 1,
            });

            expect(result.addItemToOrder.lines.length).toBe(1);
            expect(result.addItemToOrder.lines[0].quantity).toBe(1);
            expect(result.addItemToOrder.lines[0].productVariant.id).toBe('T_1');
            expect(result.addItemToOrder.lines[0].id).toBe('T_1');
            firstOrderItemId = result.addItemToOrder.lines[0].id;
        });

        it('addItemToOrder errors with an invalid productVariantId', async () => {
            try {
                await client.query(ADD_ITEM_TO_ORDER, {
                    productVariantId: 'T_999',
                    quantity: 1,
                });
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining(`No ProductVariant with the id '999' could be found`),
                );
            }
        });

        it('addItemToOrder errors with a negative quantity', async () => {
            try {
                await client.query(ADD_ITEM_TO_ORDER, {
                    productVariantId: 'T_999',
                    quantity: -3,
                });
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining(`-3 is not a valid quantity for an OrderItem`),
                );
            }
        });

        it('addItemToOrder with an existing productVariantId adds quantity to the existing OrderLine', async () => {
            const result = await client.query(ADD_ITEM_TO_ORDER, {
                productVariantId: 'T_1',
                quantity: 2,
            });

            expect(result.addItemToOrder.lines.length).toBe(1);
            expect(result.addItemToOrder.lines[0].quantity).toBe(3);
        });

        it('adjustItemQuantity adjusts the quantity', async () => {
            const result = await client.query(ADJUST_ITEM_QUENTITY, {
                orderItemId: firstOrderItemId,
                quantity: 50,
            });

            expect(result.adjustItemQuantity.lines.length).toBe(1);
            expect(result.adjustItemQuantity.lines[0].quantity).toBe(50);
        });

        it('adjustItemQuantity errors with a negative quantity', async () => {
            try {
                await client.query(ADJUST_ITEM_QUENTITY, {
                    orderItemId: firstOrderItemId,
                    quantity: -3,
                });
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining(`-3 is not a valid quantity for an OrderItem`),
                );
            }
        });

        it('adjustItemQuantity errors with an invalid orderItemId', async () => {
            try {
                await client.query(ADJUST_ITEM_QUENTITY, {
                    orderItemId: 'T_999',
                    quantity: 5,
                });
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining(`This order does not contain an OrderLine with the id 999`),
                );
            }
        });

        it('removeItemFromOrder removes the correct item', async () => {
            const result1 = await client.query(ADD_ITEM_TO_ORDER, {
                productVariantId: 'T_3',
                quantity: 3,
            });
            expect(result1.addItemToOrder.lines.length).toBe(2);
            expect(result1.addItemToOrder.lines.map(i => i.productVariant.id)).toEqual(['T_1', 'T_3']);

            const result2 = await client.query(REMOVE_ITEM_FROM_ORDER, {
                orderItemId: firstOrderItemId,
            });
            expect(result2.removeItemFromOrder.lines.length).toBe(1);
            expect(result2.removeItemFromOrder.lines.map(i => i.productVariant.id)).toEqual(['T_3']);
        });

        it('removeItemFromOrder errors with an invalid orderItemId', async () => {
            try {
                await client.query(REMOVE_ITEM_FROM_ORDER, {
                    orderItemId: 'T_999',
                });
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining(`This order does not contain an OrderLine with the id 999`),
                );
            }
        });

        it('nextOrderStates returns next valid states', async () => {
            const result = await client.query(gql`
                query {
                    nextOrderStates
                }
            `);

            expect(result.nextOrderStates).toEqual(['ArrangingPayment']);
        });

        it('transitionOrderToState throws for an invalid state', async () => {
            try {
                await client.query(TRANSITION_TO_STATE, { state: 'Completed' });
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining(`Cannot transition Order from "AddingItems" to "Completed"`),
                );
            }
        });

        it('attempting to transition to ArrangingPayment throws when Order has no Customer', async () => {
            try {
                await client.query(TRANSITION_TO_STATE, { state: 'ArrangingPayment' });
                fail('Should have thrown');
            } catch (err) {
                expect(err.message).toEqual(
                    expect.stringContaining(
                        `Cannot transition Order to the "ArrangingShipping" state without Customer details`,
                    ),
                );
            }
        });

        it('setCustomerForOrder creates a new Customer and associates it with the Order', async () => {
            const result = await client.query(SET_CUSTOMER, {
                input: {
                    emailAddress: 'test@test.com',
                    firstName: 'Test',
                    lastName: 'Person',
                },
            });

            const customer = result.setCustomerForOrder.customer;
            expect(customer.firstName).toBe('Test');
            expect(customer.lastName).toBe('Person');
            expect(customer.emailAddress).toBe('test@test.com');
            createdCustomerId = customer.id;
        });

        it('setCustomerForOrder updates the existing customer if Customer already set', async () => {
            const result = await client.query(SET_CUSTOMER, {
                input: {
                    emailAddress: 'test@test.com',
                    firstName: 'Changed',
                    lastName: 'Person',
                },
            });

            const customer = result.setCustomerForOrder.customer;
            expect(customer.firstName).toBe('Changed');
            expect(customer.lastName).toBe('Person');
            expect(customer.emailAddress).toBe('test@test.com');
            expect(customer.id).toBe(createdCustomerId);
        });

        it('can transition to ArrangingPayment once Customer has been set', async () => {
            const result = await client.query(TRANSITION_TO_STATE, { state: 'ArrangingPayment' });

            expect(result.transitionOrderToState).toEqual({ id: 'T_1', state: 'ArrangingPayment' });
        });
    });

    describe('as authenticated user', () => {
        let firstOrderItemId: string;
        let activeOrder: any;
        let authenticatedUserEmailAddress: string;
        let customers: GetCustomerList.Items[];
        const password = 'test';

        beforeAll(async () => {
            await client.asSuperAdmin();
            const result = await client.query<GetCustomerList.Query, GetCustomerList.Variables>(
                GET_CUSTOMER_LIST,
                {
                    options: {
                        take: 2,
                    },
                },
            );
            customers = result.customers.items;
            authenticatedUserEmailAddress = customers[0].emailAddress;
            await client.asUserWithCredentials(authenticatedUserEmailAddress, password);
        });

        it('activeOrder returns null before any items have been added', async () => {
            const result = await client.query(GET_ACTIVE_ORDER);
            expect(result.activeOrder).toBeNull();
        });

        it('addItemToOrder creates a new Order with an item', async () => {
            const result = await client.query(ADD_ITEM_TO_ORDER, {
                productVariantId: 'T_1',
                quantity: 1,
            });

            expect(result.addItemToOrder.lines.length).toBe(1);
            expect(result.addItemToOrder.lines[0].quantity).toBe(1);
            expect(result.addItemToOrder.lines[0].productVariant.id).toBe('T_1');
            activeOrder = result.addItemToOrder;
            firstOrderItemId = result.addItemToOrder.lines[0].id;
        });

        it('activeOrder returns order after item has been added', async () => {
            const result = await client.query(GET_ACTIVE_ORDER);
            expect(result.activeOrder.id).toBe(activeOrder.id);
            expect(result.activeOrder.state).toBe('AddingItems');
        });

        it('addItemToOrder with an existing productVariantId adds quantity to the existing OrderLine', async () => {
            const result = await client.query(ADD_ITEM_TO_ORDER, {
                productVariantId: 'T_1',
                quantity: 2,
            });

            expect(result.addItemToOrder.lines.length).toBe(1);
            expect(result.addItemToOrder.lines[0].quantity).toBe(3);
        });

        it('adjustItemQuantity adjusts the quantity', async () => {
            const result = await client.query(ADJUST_ITEM_QUENTITY, {
                orderItemId: firstOrderItemId,
                quantity: 50,
            });

            expect(result.adjustItemQuantity.lines.length).toBe(1);
            expect(result.adjustItemQuantity.lines[0].quantity).toBe(50);
        });

        it('removeItemFromOrder removes the correct item', async () => {
            const result1 = await client.query(ADD_ITEM_TO_ORDER, {
                productVariantId: 'T_3',
                quantity: 3,
            });
            expect(result1.addItemToOrder.lines.length).toBe(2);
            expect(result1.addItemToOrder.lines.map(i => i.productVariant.id)).toEqual(['T_1', 'T_3']);

            const result2 = await client.query(REMOVE_ITEM_FROM_ORDER, {
                orderItemId: firstOrderItemId,
            });
            expect(result2.removeItemFromOrder.lines.length).toBe(1);
            expect(result2.removeItemFromOrder.lines.map(i => i.productVariant.id)).toEqual(['T_3']);
        });

        it('nextOrderStates returns next valid states', async () => {
            const result = await client.query(GET_NEXT_STATES);

            expect(result.nextOrderStates).toEqual(['ArrangingPayment']);
        });

        it('logging out and back in again resumes the last active order', async () => {
            await client.asAnonymousUser();
            const result1 = await client.query(GET_ACTIVE_ORDER);
            expect(result1.activeOrder).toBeNull();

            await client.asUserWithCredentials(authenticatedUserEmailAddress, password);
            const result2 = await client.query(GET_ACTIVE_ORDER);
            expect(result2.activeOrder.id).toBe(activeOrder.id);
        });

        describe('shipping', () => {
            let shippingMethods: any;

            it('setOrderShippingAddress sets shipping address', async () => {
                const address: CreateAddressInput = {
                    fullName: 'name',
                    company: 'company',
                    streetLine1: '12 the street',
                    streetLine2: 'line 2',
                    city: 'foo',
                    province: 'bar',
                    postalCode: '123456',
                    country: 'baz',
                    phoneNumber: '4444444',
                };
                const result = await client.query(SET_SHIPPING_ADDRESS, {
                    input: address,
                });

                expect(result.setOrderShippingAddress.shippingAddress).toEqual(address);
            });

            it('eligibleShippingMethods lists shipping methods', async () => {
                const result = await client.query(GET_ELIGIBLE_SHIPPING_METHODS);

                shippingMethods = result.eligibleShippingMethods;

                expect(shippingMethods).toEqual([
                    { id: 'T_1', price: 500, description: 'Standard Shipping' },
                    { id: 'T_2', price: 1000, description: 'Express Shipping' },
                ]);
            });

            it('shipping is initially unset', async () => {
                const result = await client.query(GET_ACTIVE_ORDER);

                expect(result.activeOrder.shipping).toEqual(0);
                expect(result.activeOrder.shippingMethod).toEqual(null);
            });

            it('setOrderShippingMethod sets the shipping method', async () => {
                const result = await client.query(SET_SHIPPING_METHOD, {
                    id: shippingMethods[1].id,
                });

                const activeOrderResult = await client.query(GET_ACTIVE_ORDER);

                const order = activeOrderResult.activeOrder;

                expect(order.shipping).toBe(shippingMethods[1].price);
                expect(order.shippingMethod.id).toBe(shippingMethods[1].id);
                expect(order.shippingMethod.description).toBe(shippingMethods[1].description);
            });

            it('shipping method is preserved after adjustItemQuantity', async () => {
                const activeOrderResult = await client.query(GET_ACTIVE_ORDER);
                activeOrder = activeOrderResult.activeOrder;
                const result = await client.query(ADJUST_ITEM_QUENTITY, {
                    orderItemId: activeOrder.lines[0].id,
                    quantity: 10,
                });

                expect(result.adjustItemQuantity.shipping).toBe(shippingMethods[1].price);
                expect(result.adjustItemQuantity.shippingMethod.id).toBe(shippingMethods[1].id);
                expect(result.adjustItemQuantity.shippingMethod.description).toBe(
                    shippingMethods[1].description,
                );
            });
        });

        describe('payment', () => {
            it('attempting add a Payment throws error when in AddingItems state', async () => {
                try {
                    await client.query(ADD_PAYMENT, {
                        input: {
                            method: testPaymentMethod.code,
                            metadata: {},
                        },
                    });
                    fail('Should have thrown');
                } catch (err) {
                    expect(err.message).toEqual(
                        expect.stringContaining(
                            `A Payment may only be added when Order is in "ArrangingPayment" state`,
                        ),
                    );
                }
            });

            it('transitions to the ArrangingPayment state', async () => {
                const result = await client.query(TRANSITION_TO_STATE, { state: 'ArrangingPayment' });
                expect(result.transitionOrderToState).toEqual({
                    id: activeOrder.id,
                    state: 'ArrangingPayment',
                });
            });

            it('attempting to add an item throws error when in ArrangingPayment state', async () => {
                try {
                    const result = await client.query(ADD_ITEM_TO_ORDER, {
                        productVariantId: 'T_4',
                        quantity: 1,
                    });
                    fail('Should have thrown');
                } catch (err) {
                    expect(err.message).toEqual(
                        expect.stringContaining(
                            `Order contents may only be modified when in the "AddingItems" state`,
                        ),
                    );
                }
            });

            it('attempting to modify item quantity throws error when in ArrangingPayment state', async () => {
                try {
                    const result = await client.query(ADJUST_ITEM_QUENTITY, {
                        orderItemId: activeOrder.lines[0].id,
                        quantity: 12,
                    });
                    fail('Should have thrown');
                } catch (err) {
                    expect(err.message).toEqual(
                        expect.stringContaining(
                            `Order contents may only be modified when in the "AddingItems" state`,
                        ),
                    );
                }
            });

            it('attempting to remove an item throws error when in ArrangingPayment state', async () => {
                try {
                    const result = await client.query(REMOVE_ITEM_FROM_ORDER, {
                        orderItemId: activeOrder.lines[0].id,
                    });
                    fail('Should have thrown');
                } catch (err) {
                    expect(err.message).toEqual(
                        expect.stringContaining(
                            `Order contents may only be modified when in the "AddingItems" state`,
                        ),
                    );
                }
            });

            it('attempting to setOrderShippingMethod throws error when in ArrangingPayment state', async () => {
                const shippingMethodsResult = await client.query(GET_ELIGIBLE_SHIPPING_METHODS);
                const shippingMethods = shippingMethodsResult.eligibleShippingMethods;

                try {
                    await client.query(SET_SHIPPING_METHOD, {
                        id: shippingMethods[0].id,
                    });
                    fail('Should have thrown');
                } catch (err) {
                    expect(err.message).toEqual(
                        expect.stringContaining(
                            `Order contents may only be modified when in the "AddingItems" state`,
                        ),
                    );
                }
            });

            it('adds a declined payment', async () => {
                const result = await client.query(ADD_PAYMENT, {
                    input: {
                        method: testFailingPaymentMethod.code,
                        metadata: {
                            foo: 'bar',
                        },
                    },
                });

                const payment = result.addPaymentToOrder.payments[0];
                expect(result.addPaymentToOrder.payments.length).toBe(1);
                expect(payment.method).toBe(testFailingPaymentMethod.code);
                expect(payment.state).toBe('Declined');
                expect(payment.transactionId).toBe(null);
                expect(payment.metadata).toEqual({
                    foo: 'bar',
                });
            });

            it('adds a successful payment and transitions Order state', async () => {
                const result = await client.query(ADD_PAYMENT, {
                    input: {
                        method: testPaymentMethod.code,
                        metadata: {
                            baz: 'quux',
                        },
                    },
                });

                const payment = result.addPaymentToOrder.payments[0];
                expect(result.addPaymentToOrder.state).toBe('PaymentSettled');
                expect(result.addPaymentToOrder.active).toBe(false);
                expect(result.addPaymentToOrder.payments.length).toBe(1);
                expect(payment.method).toBe(testPaymentMethod.code);
                expect(payment.state).toBe('Settled');
                expect(payment.transactionId).toBe('12345');
                expect(payment.metadata).toEqual({
                    baz: 'quux',
                });
            });
        });

        describe('orderByCode', () => {
            describe('immediately after Order is placed', () => {
                it('works when authenticated', async () => {
                    const result = await client.query(GET_ORDER_BY_CODE, {
                        code: activeOrder.code,
                    });

                    expect(result.orderByCode.id).toBe(activeOrder.id);
                });

                it('works when anonymous', async () => {
                    await client.asAnonymousUser();
                    const result = await client.query(GET_ORDER_BY_CODE, {
                        code: activeOrder.code,
                    });

                    expect(result.orderByCode.id).toBe(activeOrder.id);
                });

                it(`throws error for another user's Order`, async () => {
                    authenticatedUserEmailAddress = customers[1].emailAddress;
                    await client.asUserWithCredentials(authenticatedUserEmailAddress, password);

                    try {
                        await client.query(GET_ORDER_BY_CODE, {
                            code: activeOrder.code,
                        });
                        fail('Should have thrown');
                    } catch (err) {
                        expect(err.message).toEqual(
                            expect.stringContaining(
                                `You are not currently authorized to perform this action`,
                            ),
                        );
                    }
                });
            });
        });
    });
});

const testPaymentMethod = new PaymentMethodHandler({
    code: 'test-payment-method',
    name: 'Test Payment Method',
    args: {},
    createPayment: (order, args, metadata) => {
        return {
            amount: order.total,
            state: 'Settled',
            transactionId: '12345',
            metadata,
        };
    },
});

const testFailingPaymentMethod = new PaymentMethodHandler({
    code: 'test-failing-payment-method',
    name: 'Test Failing Payment Method',
    args: {},
    createPayment: (order, args, metadata) => {
        return {
            amount: order.total,
            state: 'Declined',
            metadata,
        };
    },
});

const TEST_ORDER_FRAGMENT = gql`
    fragment TestOrderFragment on Order {
        id
        code
        state
        active
        lines {
            id
            quantity
            productVariant {
                id
            }
        }
        shipping
        shippingMethod {
            id
            code
            description
        }
    }
`;

const GET_ACTIVE_ORDER = gql`
    query {
        activeOrder {
            ...TestOrderFragment
        }
    }
    ${TEST_ORDER_FRAGMENT}
`;

const ADD_ITEM_TO_ORDER = gql`
    mutation AddItemToOrder($productVariantId: ID!, $quantity: Int!) {
        addItemToOrder(productVariantId: $productVariantId, quantity: $quantity) {
            ...TestOrderFragment
        }
    }
    ${TEST_ORDER_FRAGMENT}
`;

const ADJUST_ITEM_QUENTITY = gql`
    mutation AdjustItemQuantity($orderItemId: ID!, $quantity: Int!) {
        adjustItemQuantity(orderItemId: $orderItemId, quantity: $quantity) {
            ...TestOrderFragment
        }
    }
    ${TEST_ORDER_FRAGMENT}
`;

const REMOVE_ITEM_FROM_ORDER = gql`
    mutation RemoveItemFromOrder($orderItemId: ID!) {
        removeItemFromOrder(orderItemId: $orderItemId) {
            ...TestOrderFragment
        }
    }
    ${TEST_ORDER_FRAGMENT}
`;

const GET_NEXT_STATES = gql`
    query {
        nextOrderStates
    }
`;

const TRANSITION_TO_STATE = gql`
    mutation TransitionToState($state: String!) {
        transitionOrderToState(state: $state) {
            id
            state
        }
    }
`;

const GET_ELIGIBLE_SHIPPING_METHODS = gql`
    query {
        eligibleShippingMethods {
            id
            price
            description
        }
    }
`;

const SET_SHIPPING_ADDRESS = gql`
    mutation SetShippingAddress($input: CreateAddressInput!) {
        setOrderShippingAddress(input: $input) {
            shippingAddress {
                fullName
                company
                streetLine1
                streetLine2
                city
                province
                postalCode
                country
                phoneNumber
            }
        }
    }
`;

const SET_SHIPPING_METHOD = gql`
    mutation SetShippingMethod($id: ID!) {
        setOrderShippingMethod(shippingMethodId: $id) {
            shipping
            shippingMethod {
                id
                code
                description
            }
        }
    }
`;

const ADD_PAYMENT = gql`
    mutation AddPaymentToOrder($input: PaymentInput!) {
        addPaymentToOrder(input: $input) {
            ...TestOrderFragment
            payments {
                id
                transactionId
                method
                amount
                state
                metadata
            }
        }
    }
    ${TEST_ORDER_FRAGMENT}
`;

const SET_CUSTOMER = gql`
    mutation SetCustomerForOrder($input: CreateCustomerInput!) {
        setCustomerForOrder(input: $input) {
            id
            customer {
                id
                emailAddress
                firstName
                lastName
            }
        }
    }
`;

const GET_ORDER_BY_CODE = gql`
    query GetOrderByCode($code: String!) {
        orderByCode(code: $code) {
            ...TestOrderFragment
        }
    }
    ${TEST_ORDER_FRAGMENT}
`;
