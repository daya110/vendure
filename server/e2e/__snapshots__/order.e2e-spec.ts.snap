// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders as anonymous user addItemToOrder() creates a new Order with an item 1`] = `
Object {
  "id": "T_1",
  "productVariant": Object {
    "id": "T_1",
  },
  "quantity": 1,
}
`;
